# Task 10: Introduction to NumPy

Learn the basics of NumPy (arrays, etc.) and perform some array operations in multiple dimensions using NumPy.

In [2]:
# import NumPy with alias as np
import numpy as np

## 1. 1D Array

In [3]:
arr1 = np.arange(10)
print("1D NumPy array filled with integers from 0 to 9: \n", arr1)

1D NumPy array filled with integers from 0 to 9: 
 [0 1 2 3 4 5 6 7 8 9]


## 2. 2D Array

In [4]:
# Seed the random number generator for reproducibility (optional)
np.random.seed(0)

# Create a 3x3 array with random integers between 1 and 20
arr2 = np.random.randint(1, 21, size=(3, 3))

print("2D NumPy array (3x3) with random integers between 1 and 20: \n", arr2)

2D NumPy array (3x3) with random integers between 1 and 20: 
 [[13 16  1]
 [ 4  4  8]
 [10 20 19]]


## 3. 3D Array

In [5]:
# 3D Array with dimensions (2,3,4) filled with ones
arr3 = np.ones((2, 3, 4))
# print("1D NumPy array filled with integers from 0 to 9: \n", arr1)
arr3

array([[[1., 1., 1., 1.],
        [1., 1., 1., 1.],
        [1., 1., 1., 1.]],

       [[1., 1., 1., 1.],
        [1., 1., 1., 1.],
        [1., 1., 1., 1.]]])

### 4. Add two 1D arrays element wise

In [17]:
print(f"{arr1} + {arr1}")
arr = arr1 + arr1
print("Addition of two 1D arrays is: \n", arr)

[0 1 2 3 4 5 6 7 8 9] + [0 1 2 3 4 5 6 7 8 9]
Addition of two 1D arrays is: 
 [ 0  2  4  6  8 10 12 14 16 18]


### 5. Multiply two 2D arrays element wise

In [19]:
print(f"{arr2} * {arr2}")
mul_arr = arr2 * arr2
print("Multiplication of two 2D arrays is: \n", mul_arr)

[[13 16  1]
 [ 4  4  8]
 [10 20 19]] * [[13 16  1]
 [ 4  4  8]
 [10 20 19]]
Multiplication of two 2D arrays is: 
 [[169 256   1]
 [ 16  16  64]
 [100 400 361]]


### 6. Calculate the dot product of two matrices.

In [9]:
# Define two matrices
matrix_A = np.array([[1, 2],
                     [3, 4]])

matrix_B = np.array([[5, 6],
                     [7, 8]])

# calculate dot product of two matrices
dot_product = np.dot(matrix_A, matrix_B)

# print matrices and their dot product
print("Matrix A: \n", matrix_A)
print("Matrix B: \n", matrix_B)
print("Dot product of the Matrix A and Matrix B is: \n", dot_product)

Matrix A: 
 [[1 2]
 [3 4]]
Matrix B: 
 [[5 6]
 [7 8]]
Dot product of the Matrix A and Matrix B is: 
 [[19 22]
 [43 50]]


### 7. Calculating the Mean, Median & Std. Deviation of 1D array

In [24]:
# Calculate mean
mean_val = np.mean(arr1)

# Calculate median
median_val = np.median(arr1)

# Calculate std. deviation
std_dev = np.std(arr1)

# Print results
print(f"1D array: {arr1}")
print("Mean:", mean_val)
print("Median:", median_val)
print("Standard Deviation:", std_dev)


1D array: [0 1 2 3 4 5 6 7 8 9]
Mean: 4.5
Median: 4.5
Standard Deviation: 2.8722813232690143


### 8. Find the maximum and minimum values in a 2D array.

In [7]:
# Find max value in the 2D array
max_val = np.max(arr2)

# Find min. value in the 2D array
min_val = np.min(arr2)

print("2D Array: \n", arr2)
print("Maximum value in the 2D array is: ", max_val)
print("Minimum value in the 2D array is: ", min_val)

2D Array: 
 [[13 16  1]
 [ 4  4  8]
 [10 20 19]]
Maximum value in the 2D array is:  20
Minimum value in the 2D array is:  1


### 9. Generate an array of 1000 random numbers from a normal distribution with a mean of 0 and a standard deviation of 1.

In [10]:
random_numbers = np.random.normal(loc=0, scale=1, size=1000)

# print first 10 numbers to verify
print("First 10 random numbers from the normal distribution: ")
print(random_numbers[:10])

# print some statistics to verify the distribution
print("\n Mean:", np.mean(random_numbers))
print("Standard Deviation:", np.std(random_numbers))

First 10 random numbers from the normal distribution: 
[ 0.95008842 -0.15135721 -0.10321885  0.4105985   0.14404357  1.45427351
  0.76103773  0.12167502  0.44386323  0.33367433]

 Mean: -0.049970459293933185
Standard Deviation: 0.9805502122823032


### 10. Create a 2D array of shapes (5, 5) with random integers from a uniform distribution between 10 and 50.

In [11]:
# create 2D array of shape(5x5) with random integers between 10 & 50
array_2d = np.random.randint(10, 51, size=(5, 5))

# print 2D array
print("2D Array (5x5): ")
print(array_2d)

2D Array (5x5): 
[[18 25 46 11 40]
 [47 10 29 16 39]
 [13 41 39 26 13]
 [36 16 47 21 27]
 [13 30 37 38 24]]


### 11. Cumulative sum of 1D array

In [26]:
cumulative_sum = np.cumsum(arr1)

# print the 1D array and its cumulative sum
print("1D array: ", arr1)
print("Cumulative Sum:", cumulative_sum)

1D array:  [0 1 2 3 4 5 6 7 8 9]
Cumulative Sum: [ 0  1  3  6 10 15 21 28 36 45]


### 12. Compute the correlation coefficient matrix of a 2D array.

In [12]:
# considering 2D array of shape (5x5)

# pompute the correlation coefficient matrix
corre_matrix = np.corrcoef(array_2d)

# print 2D array (5x5) and the correlation coefficient matrix
print("2D Array (5x5) is:")
print(array_2d)
print("\n Correlation Coefficient Matrix:")
print(corre_matrix)

2D Array (5x5) is:
[[18 25 46 11 40]
 [47 10 29 16 39]
 [13 41 39 26 13]
 [36 16 47 21 27]
 [13 30 37 38 24]]

 Correlation Coefficient Matrix:
[[ 1.          0.24024423  0.20483802  0.55759148  0.14501973]
 [ 0.24024423  1.         -0.77644378  0.60257279 -0.74925653]
 [ 0.20483802 -0.77644378  1.         -0.04015473  0.70517806]
 [ 0.55759148  0.60257279 -0.04015473  1.         -0.0821296 ]
 [ 0.14501973 -0.74925653  0.70517806 -0.0821296   1.        ]]


This correlation coefficient matrix represents the pairwise correlations between variables in a dataset. Each value in the matrix shows the strength and direction of the linear relationship between two variables.
- For instance, the correlation of 0.55759148 between the first and fourth variables indicates a moderate positive relationship.
- while -0.77644378 between the second and third variables indicates a strong negative relationship.
- The diagonal elements are 1, as each variable is perfectly correlated with itself.

### 13. Simulate rolling a six-sided die 1000 times and count the frequency of each outcome.

In [13]:
# simulate rolling a six-sided die 1000 times
rolls = np.random.randint(1, 7, size=1000)

# count frequency of each outcome (1 through 6)
frequencies = np.bincount(rolls, minlength=7)[1:]

# print the outcomes and their frequencies
print("Outcome frequencies after rolling a six-sided die 1000 times:")
for outcome, frequency in enumerate(frequencies, start=1):
    print(f"{outcome}: {frequency}")

Outcome frequencies after rolling a six-sided die 1000 times:
1: 149
2: 165
3: 167
4: 182
5: 166
6: 171
